### PR TITLE
L86: Start new gRFC for aspect-based python bazel rules

### PR DESCRIPTION
**Pull request by mbeards**

### CODE REVIEW

File: L81-aspect-based-python-bazel-rules.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+Aspect-based Python Bazel Rules
+----
+* Author(s): Michael Beardsworth
+* Approver: a11r
```
Comment (gnossen) [inline L81-aspect-based-python-bazel-rules.md:4]: You can change the reviewer to me.

File: L81-aspect-based-python-bazel-rules.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+* Aspects `_gen_py_proto_aspect` and `_gen_py_grpc_aspect` visit `proto_library` rules to generate Python code for Protobuf and gRPC, respectively.
+* They produce custom [providers](https://docs.bazel.build/versions/main/skylark/rules.html#providers) `PyProtoInfo` and `PyGrpcInfo` that wrap a `PyInfo` provider to avoid creating spurious dependencies for Python users that interface with the `proto_library` rules through some other means.
+* The `py_proto_library` and `py_grpc_library` rules will only be responsible for collecting the `PyInfo` providers from their dependencies.
+* The `plugin` attribute must be removed from `py_proto_library` and `py_grpc_library`. Aspects require the declaration of all possible parameter values up front, so it would not be possible for the new aspects to continue supporting arbitrary plugins. (Note that the plugin feature is not used in gRPC. It was introduced to support [GAPIC](https://github.com/googleapis/gapic-generator-python), which no longer uses the feature.)
+* No behavior change should be observed by the user of `py_proto_library` or `py_grpc_library` unless they rely on the (removed) `plugin` attribute.
```
Comment (gnossen) [inline L81-aspect-based-python-bazel-rules.md:34]: In the grpc-io thread for this gRFC, this point should be bolded and near the top.

File: L81-aspect-based-python-bazel-rules.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
+Aspect-based Python Bazel Rules
+----
+* Author(s): Michael Beardsworth
+* Approver: a11r
+* Status: Draft
+* Implemented in: Bazel and Starlark
+* Last updated: [Date]
+* Discussion at: <google group thread> (filled after thread exists)
```
Comment (gnossen) [inline L81-aspect-based-python-bazel-rules.md:8]: Don't forget to create the thread and link it on this line.

File: L81-aspect-based-python-bazel-rules.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+## Background
+
+* gRPC provides build rule logic written in Starlark to support Protobuf and gRPC Python codegen with Bazel.
+* These rules (`py_proto_library` and `py_grpc_library`) each generate code from a single `proto_library`'s protos.
+* The produced Python libraries (or equivalently-functioning PyInfo providers) do not express dependencies on the generated Python code for the `proto_library`'s dependencies.
+* Users are surprised by this behavior. See [#23769](https://github.com/grpc/grpc/issues/23769) for an example bug.
+* This behavior is particularly surprising for Google-internal users. The internal versions of `py_proto_library` and `py_grpc_library` propagate Python dependencies.
+
+### Related Proposals:
+`py_proto_library` and `py_grpc_library` are implemented in [python_rules.bzl](https://github.com/grpc/grpc/blob/bazel/python_rules.bzl). There doesn't seem to be a proposal for that, though.
```
Comment (gnossen) [inline L81-aspect-based-python-bazel-rules.md:24]: > There doesn't seem to be a proposal for that, though.

To give you some background here, these two rules were originally intended to be used only by the `grpc/grpc` repo. It wasn't until later that we found that lots of folks were using them externally as well.

**Review by gnossen — APPROVED** [review]
Thanks for putting the time into writing this up. I think it's unlikely that anyone will complain, but better safe than sorry.

File: L81-aspect-based-python-bazel-rules.md (hunk 1 above)
Comment (mbeards) [reply L81-aspect-based-python-bazel-rules.md]: Done.

File: L81-aspect-based-python-bazel-rules.md (hunk 3 above)
Comment (mbeards) [reply L81-aspect-based-python-bazel-rules.md:8]: Ack. Should I do that in a follow-up before landing the PR?

File: L81-aspect-based-python-bazel-rules.md (hunk 4 above)
Comment (mbeards) [reply L81-aspect-based-python-bazel-rules.md:24]: Hyrum's Law in action! It certainly seems like there's demand - maybe it's worth considering upstreaming this to [https://github.com/bazelbuild/rules_python](https://github.com/bazelbuild/rules_python) in a future change.

File: L81-aspect-based-python-bazel-rules.md (hunk 3 above)
Comment (gnossen) [reply L81-aspect-based-python-bazel-rules.md]: Yes. The merged PR of the gRFC should include a link to the email thread. The process is:

1. Create the PR with the link not included. (like you have here)
2. Create the discussion at `grpc-io` with the link to the PR just created.
3. Update the PR to include the link to the discussion.
4. Wait 2 weeks.
5. Merge PR.

File: L81-aspect-based-python-bazel-rules.md (hunk 2 above)
Comment (mbeards) [reply L81-aspect-based-python-bazel-rules.md]: Done.

File: L81-aspect-based-python-bazel-rules.md (hunk 4 above)
Comment (gnossen) [reply L81-aspect-based-python-bazel-rules.md]: Perhaps. Several other repos (including I think `rules_python`) had similar rules but they were much less rosy about other repos using their rules. There would definitely need to be a discussion about it.

File: L81-aspect-based-python-bazel-rules.md (hunk 3 above)
Comment (mbeards) [reply L81-aspect-based-python-bazel-rules.md]: Gotcha, thanks.

File: L86-aspect-based-python-bazel-rules.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+The proposed approach addresses the open bug and corrects the dependency issues. However, it requires the removal of a (likely unused) feature of the build rules.
+
+Alternatively a set of repository rules could be introduced that allow users to inject `py_proto_library` and `py_grpc_library` implementations into the gRPC Bazel build system.
+This alternative approach would allow users to work around the dependency issues by taking on additional burden.
+The new build logic could be moved into a separate repository, or potentially upstreamed to a Bazel-owned repository.
+
+## Implementation
+
+The implementation is mostly complete by beardsworth in [#27275](https://github.com/grpc/grpc/pull/27275).
```
Comment (lidizheng) [inline L86-aspect-based-python-bazel-rules.md:46]: Before merging, please add test cases.

File: L86-aspect-based-python-bazel-rules.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  This approach corresponds to the Google-internal design for these rules.
+* The `_gen_py_proto_aspect` aspect visits `proto_library` rules to generate Python code for Protobuf.
+* The aspect produces a custom [providers](https://docs.bazel.build/versions/main/skylark/rules.html#providers) `PyProtoInfo` that wraps a `PyInfo` provider to avoid creating spurious dependencies for Python users that interface with the `proto_library` rules through some other means.
+* The `py_proto_library` and `py_grpc_library` rules will only be responsible for collecting the `PyInfo` providers from their dependencies.
+* The `plugin` attribute must be removed from `py_proto_library`. Aspects require the declaration of all possible parameter values up front, so it would not be possible for the new aspects to continue supporting arbitrary plugins. (Note that the plugin feature is not used in gRPC. It was introduced to support [GAPIC](https://github.com/googleapis/gapic-generator-python), which no longer uses the feature.)
+* No behavior change should be observed by the user of `py_proto_library` or `py_grpc_library` unless they rely on the (removed) `plugin` attribute.
```
Comment (lidizheng) [inline L86-aspect-based-python-bazel-rules.md:34]: We might want to note this in v1.41.x release note.

File: L86-aspect-based-python-bazel-rules.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+* This behavior is particularly surprising for Google-internal users. The internal versions of `py_proto_library` and `py_grpc_library` propagate Python dependencies.
+
+### Related Proposals:
+`py_proto_library` and `py_grpc_library` are implemented in [python_rules.bzl](https://github.com/grpc/grpc/blob/bazel/python_rules.bzl). There doesn't seem to be a proposal for that, though.
+
+## Proposal
+
+* We propose rebuilding the `py_proto_library` and `py_grpc_library` rules to use [aspects](https://docs.bazel.build/versions/main/skylark/aspects.html) to perform code generation steps.
+  This approach corresponds to the Google-internal design for these rules.
+* The `_gen_py_proto_aspect` aspect visits `proto_library` rules to generate Python code for Protobuf.
```
Comment (lidizheng) [inline L86-aspect-based-python-bazel-rules.md:30]: We might want to note the protobuf Python code is still generated by us instead of `proto_library`.

**Comment by mbeards** [comment]
I'm going to pause on this for a bit - there are enough open questions in-flight around import paths that I suspect a multi-step approach is going to be needed if we want to land a change like this.

File: L86-aspect-based-python-bazel-rules.md (hunk 5 above)
Comment (mbeards) [reply L86-aspect-based-python-bazel-rules.md:48]: Ack. I added a mention of testing in the Implementation section.

File: L86-aspect-based-python-bazel-rules.md (hunk 6 above)
Comment (mbeards) [reply L86-aspect-based-python-bazel-rules.md:36]: Anything additional to mention in the gRFC? Or should this be addressed in the implementation PR?

File: L86-aspect-based-python-bazel-rules.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+* The aspect produces a custom [providers](https://docs.bazel.build/versions/main/skylark/rules.html#providers) `PyProtoInfo` that wraps a `PyInfo` provider to avoid creating spurious dependencies for Python users that interface with the `proto_library` rules through some other means.
+* The `py_proto_library` and `py_grpc_library` rules will only be responsible for collecting the `PyInfo` providers from their dependencies.
+* The `plugin` attribute must be removed from `py_proto_library`. Aspects require the declaration of all possible parameter values up front, so it would not be possible for the new aspects to continue supporting arbitrary plugins. (Note that the plugin feature is not used in gRPC. It was introduced to support [GAPIC](https://github.com/googleapis/gapic-generator-python), which no longer uses the feature.)
+* To support legacy use cases where Python code is generated from protobufs in different packages [e.g. grpcio_channelz](https://github.com/grpc/grpc/blob/master/src/python/grpcio_channelz/grpc_channelz/v1/BUILD.bazel), we add an `imports` attribute to `py_proto_library` to allow users to tweak the import paths.
```
[the remark this answers is not in the document]
Comment (mbeards) [reply L86-aspect-based-python-bazel-rules.md]: Done.

[the remark this answers is not in the document]
Comment (mbeards) [reply L86-aspect-based-python-bazel-rules.md]: Sorry - did you mean specifically on the `imports` / legacy use case stuff, or the whole section?

**Review by gnossen — APPROVED** [review]

**Pull request merged**